lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -4227,6 +4284,64 @@ ParseResult FIRCircuitParser::parseTypeDecl() {
   return success();
 }
 
+// Parse a group declaration.
+ParseResult FIRCircuitParser::parseGroupDecl(CircuitOp circuit) {
+  auto baseIndent = getIndentation();
+
+  // A stack of all groups that are possibly parents of the current group.
+  SmallVector<std::pair<std::optional<unsigned>, GroupDeclOp>> groupStack;
+
+  // Parse a single group and add it to the groupStack.
+  auto parseOne = [&](Block *block) -> ParseResult {
+    auto indent = getIndentation();
+    StringRef id, convention;
+    LocWithInfo info(getToken().getLoc(), this);
+    consumeToken();
+    if (parseId(id, "expected group name") || parseGetSpelling(convention))
```

Comment:
It'd be good for these group ID's to be tracked like other top-level decls, to nicely report if they are duplicated (two groups with same name at same point in hierarchy) and especially if they conflict with module names:

```
FIRRTL version 3.1.0

circuit group:
  declgroup group, bind:

  module group:
    input x : UInt<1>
```

Probably this should be different "namespace" but since I think same as modules/etc should be managed similarly and reported during parsing?

Reply:
This will error out with `redefinition of symbol`, but the logic for checking that the `main` module exists trips first. So, the error right now isn't great, `no main module found`. I'll see about fixing/reordering this.

Reply:
Pushed an updated patch set to catch this in a better way than previously. New behavior produces:

```
Foo.fir:4:3: error: 'firrtl.declgroup' op cannot have the same name as the circuit
  declgroup group, bind:
  ^
```

A non-circuit name collision continues to error with:

```
Foo.fir:6:10: error: redefinition of symbol named 'A'
  module A:
         ^
Foo.fir:4:3: note: see existing symbol definition here
  declgroup A, bind:
  ^
```